windows/ReactNativeAsyncStorage/DBStorage.cpp

```diff
@@ -0,0 +1,451 @@
+            other.m_db = nullptr;
+            other.m_callback = nullptr;
+        }
+        Sqlite3Transaction& operator=(const Sqlite3Transaction&) = delete;
+        Sqlite3Transaction& operator=(Sqlite3Transaction&& rhs) {
+            if (this != &rhs) {
+                Commit();
+                std::swap(m_db, rhs.m_db);
+                std::swap(m_callback, rhs.m_callback);
+            }
+        }
+
+        explicit operator bool() const {
+            return m_db != nullptr;
+        }
+
+        void Rollback() {
+            if (m_db) {
+                Exec(m_db, *m_callback, u8"ROLLBACK");
+                m_db = nullptr;
+                m_callback = nullptr;
+            }
+        }
+
+        bool Commit() {
+            if (!m_db) {
+                return false;
+            }
+            auto result = Exec(m_db, *m_callback, u8"COMMIT");
+            m_db = nullptr;
+            m_callback = nullptr;
+            return result;
+        }
+
+        ~Sqlite3Transaction() {
+            Rollback();
+        }
+    };
+
+    // Appends argcount variables to prefix in a comma-separated list.
+    std::string MakeSQLiteParameterizedStatement(const char* prefix, int argCount) {
+        assert(argCount != 0);
+        std::string result(prefix);
+        result.reserve(result.size() + (argCount * 2) + 1);
+        result += '(';
+        for (int x = 0; x < argCount - 1; x++) {
+            result += "?,";
+        }
+        result += "?)";
+        return result;
+    }
+
+    // Checks if sqliteResult is SQLITE_OK. If not, reports the error via
+    // callback & returns false.
+    bool CheckSQLiteResult(sqlite3* db, const DBStorage::Callback& callback, int sqliteResult) {
+        if (sqliteResult == SQLITE_OK) {
+            return true;
+        }
+        else {
+            InvokeError(callback, sqlite3_errmsg(db));
+            return false;
+        }
+    }
+
+    using Statement = std::unique_ptr<sqlite3_stmt, decltype(&sqlite3_finalize)>;
+
+    // Creates a prepared SQLite statement. On error, returns nullptr
+    Statement PrepareStatement(sqlite3* db, DBStorage::Callback& callback, const char* stmt) {
+        sqlite3_stmt* pStmt{ nullptr };
+        if (!CheckSQLiteResult(db, callback, sqlite3_prepare_v2(db, stmt, -1, &pStmt, nullptr))) {
+            return { nullptr, sqlite3_finalize };
+        }
+        return { pStmt, &sqlite3_finalize };
+    }
+
+    // Binds the index-th variable in this prepared statement to str.
+    bool BindString(
+        sqlite3* db,
+        const DBStorage::Callback& callback,
+        const Statement& stmt,
+        int index,
+        const std::string& str) {
+        return CheckSQLiteResult(db, callback, sqlite3_bind_text(stmt.get(), index, str.c_str(), -1, SQLITE_TRANSIENT));
+    }
+
+} // namespace
+
+DBStorage::DBStorage() {
+    auto const localAppDataParh = winrt::ApplicationData::Current().LocalFolder().Path();
```

Comment:
I'll check for package identity before using it. Where should we store the file if this path is not available? I don't think UWP app has the permission to write to arbitrary locations?

Reply:
I think the DB name should be configurable at either build-time or run-time. If you want to have it be configured at runtime, you could look for a string in `Windows.ApplicationMode.Core.CoreApplication.Properties`... say, `React-Native-Community-Async-Storage-Database-Path`? Verbose enough? :) On UWP, if there's no such value, you'd fallback to `ApplicationData`. For Win32, you'd want to error out.

Build-time configuration would be simpler, but it would preclude distributing asyncStorage as a DLL. I forget if that's a goal or not...